fortanix-vme/fortanix-vme-runner/src/lib.rs

```diff
@@ -89,14 +89,98 @@ impl Listener {
+            // According to the `Read` threat documentation, reading 0 bytes
+            // indicates that the connection has been shutdown correctly. So we
+            // close the proxy service
+            // https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
+            match Server::transfer_data(src, src_name, dst, dst_name) {
+                Ok(n) if n == 0 => false,
+                Ok(_)           => true,
+                Err(_)          => false,
+            }
+        }
+        let remote = self.tcp_stream.as_raw_fd();
+        let proxy = self.vsock_stream.as_raw_fd();
+
+        let mut read_set = FdSet::new();
+        read_set.insert(remote);
+        read_set.insert(proxy);
+
+        if let Ok(_num) = select(None, Some(&mut read_set), None, None, None) {
+            if read_set.contains(remote) {
+                return exchange(&mut self.tcp_stream, &self.remote_name, &mut self.vsock_stream, "proxy");
+            }
+            if read_set.contains(proxy) {
+                return exchange(&mut self.vsock_stream, "proxy", &mut self.tcp_stream, &self.remote_name);
+            }
+        }
+        true
+    }
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+struct ConnectionKey {
+    enclave: VsockAddr,
+    runner: VsockAddr,
+}
+
+impl ConnectionKey {
+    fn from_vsock_stream(runner_enclave: &VsockStream<Std>) -> Self {
+        let runner_cid = runner_enclave.local_addr().unwrap().cid();
+        let runner_port = runner_enclave.local_addr().unwrap().port();
+        let enclave_cid = runner_enclave.peer_addr().unwrap().cid();
+        let enclave_port = runner_enclave.peer_addr().unwrap().port();
+        Self::connection_key(enclave_cid, enclave_port, runner_cid, runner_port)
+    }
+
+    fn from_addresses(enclave: VsockAddr, runner: VsockAddr) -> Self {
+        ConnectionKey {
+            enclave,
+            runner,
+        }
+    }
+
+    fn connection_key(enclave_cid: u32, enclave_port: u32, runner_cid: u32, runner_port: u32) -> Self {
+        let enclave = VsockAddr::new(enclave_cid, enclave_port);
+        let runner = VsockAddr::new(runner_cid, runner_port);
+        Self::from_addresses(enclave, runner)
+    }
+}
+
 pub struct Server {
     command_listener: Mutex<VsockListener>,
     /// Tracks information about TCP sockets that are currently listening for new connections. For
     /// every TCP listener socket in the runner, there is a vsock listener socket in the enclave.
     /// When the enclave instructs to accept a new connection, the runner accepts a new TCP
     /// connection. It then locates the ListenerInfo and finds the information it needs to set up a
     /// new vsock connection to the enclave
-    listeners: Mutex<FnvHashMap<VsockAddr, Arc<Mutex<Listener>>>>,
+    listeners: RwLock<FnvHashMap<VsockAddr, Arc<Mutex<Listener>>>>,
+    connections: RwLock<FnvHashMap<ConnectionKey, Arc<Mutex<Connection>>>>,
```

Comment:
Just an idea that I haven't thought through completely: I think we could get rid of both the outer `RwLock` and the inner `Mutex` by assigning ownership of connections to a specific thread. What I mean is that, instead of giving multiple threads (non-mutable) access to the entire `Arc<Server>`, we could give the thread that handles the proxying ownership of just the `HashMap` containing the connections. Then that thread wouldn't need any locking for accessing the `HashMap` or the `Connection`s contained within. Of course, that thread would have be responsible for adding connections to the `HashMap` as well. I think that could be achieved by sending `Connection`s across a channel from the command thread to the proxying thread(s).